bpf/flows.c

```diff
@@ -91,58 +91,48 @@ static inline int flow_monitor(struct __sk_buff *skb, u8 direction) {
     }
     // TODO: we need to add spinlock here when we deprecate versions prior to 5.1, or provide
     // a spinlocked alternative version and use it selectively https://lwn.net/Articles/779120/
-    flow_metrics *aggregate_flow = (flow_metrics *)bpf_map_lookup_elem(&aggregated_flows, &id);
+    flow_value *aggregate_flow = (flow_value *)bpf_map_lookup_elem(&aggregated_flows, &id);
     if (aggregate_flow != NULL) {
-        aggregate_flow->packets += 1;
-        aggregate_flow->bytes += skb->len;
-        aggregate_flow->end_mono_time_ts = pkt.current_ts;
+        bpf_spin_lock(&aggregate_flow->lock);
+        aggregate_flow->flow_metrics.packets += 1;
+        aggregate_flow->flow_metrics.bytes += skb->len;
+        aggregate_flow->flow_metrics.end_mono_time_ts = pkt.current_ts;
         // it might happen that start_mono_time hasn't been set due to
         // the way percpu hashmap deal with concurrent map entries
-        if (aggregate_flow->start_mono_time_ts == 0) {
-            aggregate_flow->start_mono_time_ts = pkt.current_ts;
-        }
-        aggregate_flow->flags |= pkt.flags;
-        aggregate_flow->dscp = pkt.dscp;
-        aggregate_flow->dns_record.id = pkt.dns_id;
-        aggregate_flow->dns_record.flags = pkt.dns_flags;
-        aggregate_flow->dns_record.latency = pkt.dns_latency;
-        aggregate_flow->dns_record.errno = dns_errno;
-        long ret = bpf_map_update_elem(&aggregated_flows, &id, aggregate_flow, BPF_ANY);
-        if (ret != 0) {
-            // usually error -16 (-EBUSY) is printed here.
-            // In this case, the flow is dropped, as submitting it to the ringbuffer would cause
-            // a duplicated UNION of flows (two different flows with partial aggregation of the same packets),
-            // which can't be deduplicated.
-            // other possible values https://chromium.googlesource.com/chromiumos/docs/+/master/constants/errnos.md
-            if (trace_messages) {
-                bpf_printk("error updating flow %d\n", ret);
-            }
-            // Update global counter for hashmap update errors
-            increase_counter(HASHMAP_FLOWS_DROPPED);
+        if (aggregate_flow->flow_metrics.start_mono_time_ts == 0) {
+            aggregate_flow->flow_metrics.start_mono_time_ts = pkt.current_ts;
         }
+        aggregate_flow->flow_metrics.flags |= pkt.flags;
+        aggregate_flow->flow_metrics.dscp = pkt.dscp;
+        aggregate_flow->flow_metrics.dns_record.id = pkt.dns_id;
+        aggregate_flow->flow_metrics.dns_record.flags = pkt.dns_flags;
+        aggregate_flow->flow_metrics.dns_record.latency = pkt.dns_latency;
+        aggregate_flow->flow_metrics.dns_record.errno = dns_errno;
+        bpf_spin_unlock(&aggregate_flow->lock);
     } else {
         // Key does not exist in the map, and will need to create a new entry.
-        u64 rtt = 0;
```

Comment:
@msherif1234 I don't understand the rationale for this code. We're setting default RTT for TCP flows, how could that be accurate? I don't see how this wouldn't produce fake data.

Reply:
this the min value of ythe hook isn't fire IIRC it was a define in the kernel and I believe we doc that as the default min RTT, when I recall more I will update here